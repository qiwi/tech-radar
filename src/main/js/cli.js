#!/usr/bin/env node

import meow from 'meow'
import {generateTechRadar} from './index.js'

const cli = meow(
  `
    Usage:
      techradar  --csv-path /path/to/csv --out-dir /radar --versions 1.00 pathPrefix js
    Options
      --csvPath
      --outDir
      --version
      --pathPrefix
`,
  {
    importMeta: import.meta,
    flags: {
      csvPath: {
        type: 'string',
        isRequired: true,
      },
      outDir: {
        type: 'string',
        isRequired: true,
      },
      version: {
        type: 'string',
      },
      pathPrefix: {
        type: 'string',
      },
    },
  },
)

generateTechRadar(cli.flags)
