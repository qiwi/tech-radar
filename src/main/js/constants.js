import path from 'path'

export const langAndFw = 'languages-and-frameworks'
export const platforms = 'platforms'
export const tools = 'tools'
export const techniques = 'techniques'

export const tplDir = path.resolve('src/main/tpl')
export const tempDir = 'radar-temp'

export const radarSchema = {
  type: 'object',
  properties: {
    data: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: { type: 'string' },
          quadrant: { type: 'string' },
          ring: { type: 'string' },
          description: { type: ['string', 'null'] },
          moved: { type: ['string', 'integer', 'null'] },
        },
        required: ['name', 'quadrant', 'ring', 'description'],
      },
    },
    meta: {
      type: 'object',
      properties: {
        title: { type: 'string' },
        date: { type: 'string' },
        legend: { type: ['string', 'null'] },
      },
      required: ['title', 'date'],
    },
  },
}
export const settings = {
  svg_id: 'radar',
  width: 1450,
  height: 1100,
  colors: {
    background: '#fff',
    grid: '#bbb',
    inactive: '#ddd',
  },
  rings: [
    { name: 'ADOPT', color: '#93c47d', id: 'adopt' },
    { name: 'TRIAL', color: '#93d2c2', id: 'trial' },
    { name: 'ASSESS', color: '#fbdb84', id: 'assess' },
    { name: 'HOLD', color: '#efafa9', id: 'hold' },
  ],
  print_layout: true,
}
