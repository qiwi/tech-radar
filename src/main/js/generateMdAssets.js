import parse from 'csv-parse/lib/sync.js'
import fs from 'fs'
import fsExtra from 'fs-extra'
import path from 'path'

const tplPath = path.resolve('src/main/tpl')

export function generatePath({ name, quadrant, tempDirResolved }) {
  const entryMdName = name + '.md'
  return path.join(
    tempDirResolved,
    '/entries',
    quadrant.toLowerCase(),
    entryMdName,
  )
}

export function generateMd({ ring, description }) {
  return `---
ring: ${ring.toLowerCase()}
---
${description}`
}

export const generateMdAssets = ({ csvPath, tempDir }) => {
  const tempDirResolved = path.resolve(tempDir)
  const csvPathResolved = path.resolve(csvPath)
  fsExtra.copySync(tplPath, tempDirResolved)
  const radarData = fs.readFileSync(csvPathResolved)
  const records = parse(radarData, { columns: true })

  records.forEach(({ name, quadrant, ring, description }) => {
    const entryFilePath = generatePath({ name, quadrant, tempDirResolved })
    const content = generateMd({ ring, description })
    fs.writeFileSync(entryFilePath, content)
  })
}
