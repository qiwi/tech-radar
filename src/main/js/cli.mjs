#!/usr/bin/env node

import meow from 'meow'
import {run} from './index.js'

export const cli = meow(
  `
    Usage:
      techradar  --input /path/to/csv --output /radar --basePrefix tech-radar --autoscope true --nav-page true --nav-title title --nav-footer footer
    Options
      --input
      --output
      --basePrefix
      --autoscope
      --nav-page
      --nav-title
      --nav-footer
      --temp
`,
  {
    importMeta: import.meta,
    flags: {
      input: {
        type: 'string',
        isRequired: true,
      },
      output: {
        type: 'string',
        isRequired: true,
      },
      basePrefix: {
        type: 'string',
      },
      cwd: {
        type: 'string',
      },
      autoscope: {
        type: 'boolean'
      },
      navPage: {
        type: 'boolean'
      },
      navTitle: {
        type: 'string'
      },
      navFooter: {
        type: 'string'
      },
      temp: {
        type: 'string'
      }
    },
  },
)

run(cli.flags)
