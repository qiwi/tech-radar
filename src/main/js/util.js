// import findCacheDir from 'find-cache-dir'
import crypto from 'crypto'
import fs from 'fs'
import { ensureDirSync } from 'fs-extra'
import path from 'path'

import { settings } from './constants.js'

export const reverse = (arr) => {
  const _arr = [...arr]
  _arr.reverse()

  return _arr
}

export const makeUniq = (arr) => {
  const counters = {}
  const getCount = (k) => (counters[k] = (counters[k] || 0) + 1)

  return arr.map((k) => {
    const count = getCount(k)

    return count === 1 ? k : `${k}-${count}`
  })
}

export const ensureDir = (dir) => {
  ensureDirSync(dir)

  return dir
}
/**
 *  return unique name temp directory
 * @param cwd
 * @param temp
 * @returns {string}
 */
export const getTemp = (cwd, temp) => {
  if (temp) {
    return ensureDir(path.resolve(temp))
  }

  const id = crypto.randomBytes(16).toString('hex')
  const tempDir = id
  return ensureDir(tempDir)
}
export const getQuadrant = (quadrant, doc) => {
  const lowQuadrant = quadrant.toLowerCase()
  return doc.quadrantAliases[lowQuadrant] || lowQuadrant
}

export const writeSettings = (doc, output) => {
  const quadrants = []
  quadrants.push({ name: doc.quadrantTitles.q1 || 'Q1', id: 'q1' })
  quadrants.push({ name: doc.quadrantTitles.q2 || 'Q2', id: 'q2' })
  quadrants.push({ name: doc.quadrantTitles.q3 || 'Q3', id: 'q3' })
  quadrants.push({ name: doc.quadrantTitles.q4 || 'Q4', id: 'q4' })

  const settins = {}
  Object.assign(settins, settings)
  settins.quadrants = quadrants
  const settingsPath = path.join(output, '_data/settins.json')
  fs.writeFileSync(settingsPath, JSON.stringify(settins))
}

export const normalizeCsv = (fileContent) =>
  fileContent
    .split(/("[^"]+")/g)
    .map((item) => {
      if (item[0] === '"') return item
      return item
        .split(',')
        .map((i) => i.replace(/^ *| *$/gm, ''))
        .join(',')
    })
    .join('')

export const getDirs = (sources) => {
  const _a = sources.map((item) => item.split(path.sep))
  while (true) {
    if (_a.length === 0 || _a.length === 1) return
    const checkArray = _a.every((item) => {
      return item[0] === _a[0][0]
    })
    if (checkArray) {
      _a.forEach((item) => {
        item.shift()
        return item
      })
    } else {
      return _a.map((item) => item.join('-'))
    }
  }
}

export const sortContextsByDate = (a, b) =>
  a.date > b.date ? -1 : a.date === b.date ? 0 : 1
