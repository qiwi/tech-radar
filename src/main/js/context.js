import { uniq } from 'lodash-es'
import path from 'path'

import { read } from './reader.js'
import { getDirs } from './util.js'

/**
 * add param file to context
 * @param sources
 * @returns contexts
 */
export const init = (sources) => sources.map((s) => ({ file: s }))
/**
 * add param data to context
 * @param contexts
 * @returns contexts
 */
export const readFiles = (contexts) =>
  contexts.map((c) => ({
    ...c,
    data: read(c.file),
  }))
/**
 * add param base to context
 * @param contexts
 * @returns contexts
 */
export const resolveBases = (contexts) => {
  const dates = contexts.map(({ data }) => data.meta.date)
  if (contexts.length === 1)
    return [
      {
        ...contexts[0],
        date: contexts[0].data.meta.date,
        base: contexts[0].data.meta.date,
      },
    ]
  const filesWithNoExt = contexts.map(({ file }, i) =>
    path.join(file.slice(0, -path.basename(file).length), dates[i]),
  )

  if (uniq(filesWithNoExt).length !== filesWithNoExt.length) {
    throw new Error('same dates in one directory')
  }
  return getDirs(filesWithNoExt).map((base, i) => ({
    ...contexts[i],
    date: dates[i],
    base,
  }))
}
/**
 * sort contexts with path and date
 * @param contexts
 * @returns contexts
 */
export const sortContexts = (contexts) => {
  if (contexts.length === 1) return contexts
  return contexts.sort((a, b) => {
    if (path.dirname(a.file) > path.dirname(b.file)) return 1
    if (path.dirname(a.file) < path.dirname(b.file)) return -1
    if (a.date > b.date) return 1
    if (a.date < b.date) return -1
    return 0
  })
}
