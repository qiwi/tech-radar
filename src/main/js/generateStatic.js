import Eleventy from '@11ty/eleventy'
import fs from 'fs'
import fsExtra from 'fs-extra'
import { uniq } from 'lodash-es'
import path, { dirname } from 'path'
import { fileURLToPath } from 'url'

import {
  defNavFooter,
  defNavTitle,
  radarSchema,
  tempDir,
  tplNavPage,
} from './constants.js'
import { genMdAssets } from './generateMdAssets.js'
import { sortContextsByDate, writeSettings } from './util.js'
import { validate } from './validator.js'

const __dirname = dirname(fileURLToPath(import.meta.url))

/**
 * generate static sites from array radarDocument
 * @param contexts
 * @param _output
 * @param basePrefix
 */
export const generateStatics = async (contexts, _output, basePrefix) => {
  const statics = await contexts.reduce(async (_r, context) => {
    const _m = await _r
    const { data, base } = context
    if (!validate(data, radarSchema) || Object.keys(data).length === 0)
      return context

    const temp = tempDir
    const output = base ? path.join(_output, base) : _output
    const pathPrefix = basePrefix ? basePrefix + '/' + base : undefined

    global._11ty_ = {
      date: data.meta.date,
      title: data.meta.title,
      output,
      temp,
      pathPrefix,
    }

    try {
      genMdAssets(data, temp)
      writeSettings(data, temp)
      await genEleventy(temp, output)
    } catch (err) {
      console.error('genStatics', err)
    }
    return [..._m, output]
  }, [])
  console.log('statics=', statics)
  return contexts.map((context) => ({
    ...context,
    prefix: basePrefix ? basePrefix + '/' + context.base : context.base,
  }))
}

/**
 * generate static site with using 11ty
 * @param temp
 * @param output
 */
export const genEleventy = async (temp, output) => {
  const elev = new Eleventy(temp, output)
  elev.setConfigPathOverride('src/main/js/11ty/.eleventy.cjs')
  await elev.init()
  await elev.write()
  fsExtra.removeSync(temp)
}

export const genNavigationPage = (
  contexts,
  output,
  navPage,
  input,
  navTitle,
  navFooter,
) => {
  if (!navPage) return
  if (!navTitle) navTitle = defNavTitle
  if (!navFooter) navFooter = defNavFooter

  const dirs = contexts.map(({ file }) => path.dirname(file))
  const links = uniq(dirs.map((dir) => dir.split('/').pop())).reduce(
    (r, dir) => ({ ...r, [dir]: [] }),
    {},
  )

  contexts.sort(sortContextsByDate).forEach((context) => {
    const keys = Object.keys(links)
    const key =
      keys.length === 1
        ? keys[0]
        : keys.find((dir) => dir === context.base.split('/')[0])
    if (!key) return
    const link = `<a class="link" href=${context.base}> ${context.data.meta.date} </a>`
    links[key].push(link)
  })

  const li = (link) => `<li>${link}</li>`
  const contentPage = Object.keys(links).map(
    (key) => `<div class="tile">
<h2>${key}</h2>
<ul>
  ${links[key].map(li).join('\n ')}
</ul>
</div>`,
  )

  // Main nav page
  fsExtra.copySync(tplNavPage, output)
  const tplHtml = fs.readFileSync(path.join(output, 'index.html'), 'utf8') // eslint-disable-line sonarjs/no-duplicate-string
  const html = tplHtml
    .replace('#nav_page-content', contentPage.join('\n'))
    .replace('#nav_page-title', navTitle)
    .replace('#nav_page-footer', navFooter)
  fs.writeFileSync(path.join(output, 'index.html'), html)

  // Latest scoped radars aliases
  const redirect = fs.readFileSync(path.resolve(__dirname, '../redirect-page/index.html'), 'utf8')
  Object.entries(contexts.reduce((m, {base, date}) => {
    const scope = path.dirname(base)
    const prev = m[scope]

    if (!prev || prev < date) {
      m[scope] = date
    }
    return m
  }, {})).forEach(([scope, date]) => {
    fs.writeFileSync(path.join(output, scope, 'index.html'), redirect.replace('###', date))
  })
}
