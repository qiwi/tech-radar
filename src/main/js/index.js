import fse from 'fs-extra'
import path from 'node:path'

import { tplDir } from './constants.js'
import { genNavPage, genRadars, genRedirects } from './generator/index.js'
import { getSources, parse } from './parser/index.js'
import { getDirs, tempDir } from './util.js'

/**
 * @description
 * Generate static sites from csv/json/yml radar declarations
 *
 * @func
 * @param {Object} options
 * @param {string} options.input globby pattern for input files
 * @param {string} options.output output directory
 * @param {string} options.cwd current working directory
 * @param {string?} options.basePrefix web app root level prefix
 * @param {boolean?} options.autoscope consider same-scoped files as subversions of a single radar
 * @param {boolean?} options.navPage Generate navigation page
 * @param {string?} options.navTitle Nav page title
 * @param {string?} options.navFooter Nav page footer
 * @param {string?} options.temp Temp directory
 *
 * @return {Promise<void>}
 */
export const run = async (options) => {
  const ctx = await getContext(options)

  return readSources(ctx)
    .then(parseRadars)
    .then(sortRadars)
    .then(resolveMoves)
    .then(renderRadars)
    .finally(() => cleanTemp(ctx))
}

const getContext = async ({
  input,
  output,
  cwd = process.cwd(),
  basePrefix = '/',
  autoscope = false,
  navPage = false,
  navTitle,
  navFooter,
  temp,
  templates,
  renderSettings,
} = {}) => {
  const ctx = {
    input,
    output: path.resolve(cwd, output),
    cwd,
    basePrefix,
    autoscope,
    navPage,
    navTitle,
    navFooter,
    temp: temp || (await tempDir()),
    templates,
    renderSettings,
  }

  ctx.ctx = ctx // context self-ref to simplify pipelining

  return ctx
}

const readSources = async ({ ctx, cwd, input }) => {
  ctx.sources = await getSources(input, cwd)
  ctx.scopes = getDirs(ctx.sources).map(path.dirname)
  return ctx
}

const parseRadars = async ({ ctx, sources, scopes }) => {
  ctx.radars = await Promise.all(
    sources.map(async (file, i) => {
      const document = await parse(file)

      return {
        document,
        source: file,
        scope: scopes[i],
        date: document.meta.date,
        title: document.meta.title,
      }
    }),
  )

  return ctx
}

const renderRadars = async ({ ctx, output }) => {
  await genRadars(ctx)
  await genNavPage(ctx)
  await genRedirects(ctx)
  await fse.copy(path.join(tplDir, 'assets'), output) // shared static assets

  // console.log('radars', radars)
  // console.log('radar', JSON.stringify(radars[3], null, 2))
  return ctx
}

const resolveMoves = async ({ ctx, radars, autoscope }) => {
  if (!autoscope) {
    return ctx
  }

  const rings = {
    hold: 0,
    assess: 1,
    trial: 2,
    adopt: 3,
  }

  const getRingWeight = (ring) => rings[ring.toLowerCase()]

  radars.forEach(({ document: { data }, scope }, i) => {
    data.forEach((entry) => {
      const { name, ring, moved } = entry
      const lowerName = name.toLowerCase()
      const prevRadar = radars[i + 1] // NOTE sorted by desc date
      const prevEntry =
        prevRadar &&
        prevRadar.scope === scope &&
        prevRadar.document.data.find(
          ({ name: _name }) => _name.toLowerCase() === lowerName,
        )

      entry.moved = prevEntry
        ? Math.sign(getRingWeight(ring) - getRingWeight(prevEntry.ring))
        : moved || 0
    })
  })

  return ctx
}

const sortRadars = async ({ ctx, radars }) => {
  radars.sort((a, b) => {
    if (path.dirname(a.source) > path.dirname(b.source)) return 1
    if (path.dirname(a.source) < path.dirname(b.source)) return -1

    return Math.sign(Date.parse(b.date) - Date.parse(a.date))
  })

  return ctx
}

const cleanTemp = async ({ ctx, temp }) => {
  await fse.remove(temp)
  return ctx
}
