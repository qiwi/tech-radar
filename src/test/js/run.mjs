import fse from 'fs-extra'
import path from 'path'
import {fileURLToPath} from 'url'

import { run } from '../../main/js/index.js'
import { tempDir } from '../../main/js/util.js'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const out = path.resolve(process.cwd(), 'test')
const input = path.join(__dirname, '../stub/data/**')

export const getFileStruct = (dir, result = []) => {
  fse.readdirSync(dir).forEach((elem) => {
    const elemPath = dir + '/' + elem
    const stat = fse.statSync(elemPath)
    if (stat.isDirectory()) {
      result = [...getFileStruct(elemPath, result)]
    } else {
      result.push(elemPath)
    }
  })
  return result
}

afterAll(async () => await fse.remove(out))

describe('generate 11ty app', () => {
  it('from .csv file', async () => {
    const csvPath = path.join(__dirname, '../stub/test.csv')
    const output = await tempDir(out)

    await run({ input: csvPath, output})
    const normalizedFileStruct = getFileStruct('test')
    expect(normalizedFileStruct).toMatchSnapshot()
  })

  it('from multiple files with the same date', async () => {
    const output = await tempDir(out)
    await run({ input: 'src/test/stub/test.{csv,json,yaml}', output })
    const normalizedFileStruct = getFileStruct('test')
    expect(normalizedFileStruct).toMatchSnapshot()
  })

  it('from multiple files', async () => {
    const output = await tempDir(out)
    await run({ input: 'src/test/stub/test.{csv,json}', output })
    const normalizedFileStruct = getFileStruct('test')
    expect(normalizedFileStruct).toMatchSnapshot()
  })

  it('from .json file', async () => {
    const input = path.join(__dirname, '../stub/test.json')
    const output = await tempDir(out)

    await run({ input, output })
    const normalizedFileStruct = getFileStruct('test')
    expect(normalizedFileStruct).toMatchSnapshot()
  })

  it('from .yml file', async () => {
    const input = path.join(__dirname, '../stub/test.yml')
    const output = await tempDir(out)

    await run({ input, output })
    const normalizedFileStruct = getFileStruct('test')
    expect(normalizedFileStruct).toMatchSnapshot()
  })


  it('generate navigation page', async () => {
    const output = await tempDir(out)
    await run({ input, output, navPage: true })
    expect(fse.readFileSync(path.join(output, 'index.html'),'utf8')).toMatchSnapshot()
  })

  it('generate navigation page title', async () => {
    const output = await tempDir(out)
    await run({
      input,
      output,
      navPage: true,
      navTitle: 'title',
    })
    expect(fse.readFileSync(path.join(output, 'index.html'),'utf8')).toMatchSnapshot()
  })

  it('generate custom navigation page: title, footer', async () => {
    const output = await tempDir(out)
    await run({
      input,
      output,
      navTitle: 'test',
      navPage: true,
      templates: path.join(__dirname, '../tpl')
    })
    expect(fse.readFileSync(path.join(output, 'index.html'),'utf8')).toMatchSnapshot()
  })

  fit('generate navigation page from custom templates', async () => {
    const output = await tempDir(out)
    await run({
      input,
      output,
      templates: path.join(__dirname, '../tpl'),
      navPage: true,
    })
    expect(fse.readFileSync(path.join(output, 'index.html'),'utf8')).toMatchSnapshot()
  })

  it('generate navigation page by data2', async () => {
    const output = await tempDir(out)
    const input = 'src/test/stub/data2/**'

    await run({
      input,
      output,
      navPage: true,
      autoscope: true
    })
    expect(fse.readFileSync(path.join(output, 'index.html'),'utf8')).toMatchSnapshot()
  })
})
