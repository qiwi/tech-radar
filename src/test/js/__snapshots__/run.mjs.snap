// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate 11ty app from .csv file 1`] = `
Array [
  "favicon.ico",
  "index.html",
  "radar.css",
  "radar.js",
  "2021-06-18/index.html",
  "2021-06-18/table/index.html",
  "2021-06-18/entries/q1/TypeScript/index.html",
  "2021-06-18/entries/q2/Nodejs/index.html",
  "2021-06-18/entries/q3/codeclimate/index.html",
  "2021-06-18/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
]
`;

exports[`generate 11ty app from .json file 1`] = `
Array [
  "favicon.ico",
  "index.html",
  "radar.css",
  "radar.js",
  "2021-06-12/index.html",
  "2021-06-12/table/index.html",
  "2021-06-12/entries/q1/TypeScript/index.html",
  "2021-06-12/entries/q2/Nodejs/index.html",
  "2021-06-12/entries/q3/codeclimate/index.html",
  "2021-06-12/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
]
`;

exports[`generate 11ty app from .yml file 1`] = `
Array [
  "favicon.ico",
  "index.html",
  "radar.css",
  "radar.js",
  "2021-06-11/index.html",
  "2021-06-11/table/index.html",
  "2021-06-11/entries/q1/TypeScript/index.html",
  "2021-06-11/entries/q2/Nodejs/index.html",
  "2021-06-11/entries/q3/codeclimate/index.html",
  "2021-06-11/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
]
`;

exports[`generate 11ty app from multiple files 1`] = `
Array [
  "favicon.ico",
  "index.html",
  "radar.css",
  "radar.js",
  "2021-06-12/index.html",
  "2021-06-18/index.html",
  "2021-06-12/table/index.html",
  "2021-06-18/table/index.html",
  "2021-06-12/entries/q1/TypeScript/index.html",
  "2021-06-12/entries/q2/Nodejs/index.html",
  "2021-06-12/entries/q3/codeclimate/index.html",
  "2021-06-12/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
  "2021-06-18/entries/q1/TypeScript/index.html",
  "2021-06-18/entries/q2/Nodejs/index.html",
  "2021-06-18/entries/q3/codeclimate/index.html",
  "2021-06-18/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
]
`;

exports[`generate 11ty app from multiple files with the same date 1`] = `
Array [
  "favicon.ico",
  "index.html",
  "radar.css",
  "radar.js",
  "2021-06-12/index.html",
  "2021-06-18/index.html",
  "2021-06-12/table/index.html",
  "2021-06-18/table/index.html",
  "2021-06-12/entries/q1/TypeScript/index.html",
  "2021-06-12/entries/q2/Nodejs/index.html",
  "2021-06-12/entries/q3/codeclimate/index.html",
  "2021-06-12/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
  "2021-06-18/entries/q1/TypeScript/index.html",
  "2021-06-18/entries/q2/Nodejs/index.html",
  "2021-06-18/entries/q3/codeclimate/index.html",
  "2021-06-18/entries/q4/–ì–µ–∫—Å–∞–≥–æ–Ω–∞–ª—å–Ω–∞—è –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–∞/index.html",
]
`;

exports[`generate 11ty app generate custom navigation page: title, footer 1`] = `
"<!DOCTYPE html><head><meta charset=\\"utf-8\\"><link rel=\\"stylesheet\\" href=\\"index.css\\"><title>test</title><style>body {
          display: flex;
          flex-direction: column;
          align-items: center;
          height: 100vh;
          margin: 0;
      }
      ul {
          list-style-type: none;
          padding: 0px;
      }
      .link {
          color: black;
          text-decoration: none;
      }

      li a {
          min-width: 150px;
          padding: 3px 10px;
          border-radius: 20px;
          display: inline-block;
          font-size: 95%;
      }

      li a:hover {
          background-color: #eaeef8;
      }
      * {
          font-family: \\"Helvetiva, Museo Sans\\", sans-serif;
          color: #333;

          line-height: 1.6333;
      }
      main {
          display: flex;
          justify-content: center;
          height: 100%;
          min-width: 400px;
      }
      .wrapper {
          display: flex;
          flex-wrap: wrap;
          width: 90%;
          flex: 1;
      }
      .tile{
          text-align: center;
          margin: 20px;
      }
      footer {
          text-align: center;
      }</style></head><body><h1>test</h1><main><div class=\\"wrapper\\"><div class=\\"tile\\"><h2>.</h2><ul><li><a class=\\"link\\" href=\\"./2021-06-12\\">2021-06-12</a></li><li><a class=\\"link\\" href=\\"./2021-06-11\\">2021-06-11</a></li></ul></div></div></main><footer>foobarbaz</footer></body>"
`;

exports[`generate 11ty app generate navigation page 1`] = `
"<!DOCTYPE html><head><meta charset=\\"utf-8\\"><link rel=\\"stylesheet\\" href=\\"index.css\\"><title>üì° Tech radars</title><style>body {
          display: flex;
          flex-direction: column;
          align-items: center;
          height: 100vh;
          margin: 0;
      }
      ul {
          list-style-type: none;
          padding: 0px;
      }
      .link {
          color: black;
          text-decoration: none;
      }

      li a {
          min-width: 150px;
          padding: 3px 10px;
          border-radius: 20px;
          display: inline-block;
          font-size: 95%;
      }

      li a:hover {
          background-color: #eaeef8;
      }
      * {
          font-family: \\"Helvetiva, Museo Sans\\", sans-serif;
          color: #333;

          line-height: 1.6333;
      }
      main {
          display: flex;
          justify-content: center;
          height: 100%;
          min-width: 400px;
      }
      .wrapper {
          display: flex;
          flex-wrap: wrap;
          width: 90%;
          flex: 1;
      }
      .tile{
          text-align: center;
          margin: 20px;
      }
      footer {
          text-align: center;
      }</style></head><body><h1>üì° Tech radars</h1><main><div class=\\"wrapper\\"><div class=\\"tile\\"><h2>.</h2><ul><li><a class=\\"link\\" href=\\"./2021-06-12\\">2021-06-12</a></li><li><a class=\\"link\\" href=\\"./2021-06-11\\">2021-06-11</a></li></ul></div></div></main><footer><div style=\\"width: 100%; text-align: center; margin: 30px 0 10px 0; font-size: 80%\\"><a href=\\"https://github.com/qiwi/tech-radar\\">Tech-radar generator</a>. Based on <a href=\\"https://github.com/zalando/tech-radar\\">zalando/tech-radar</a>. Boosted with <a href=\\"https://github.com/11ty/eleventy/\\">11ty</a><br><a class=\\"link\\" href=\\"https://github.com/qiwi\\">QIWI ‚ù§Ô∏è Open Source</a></div></footer></body>"
`;

exports[`generate 11ty app generate navigation page by data2 1`] = `
"<!DOCTYPE html><head><meta charset=\\"utf-8\\"><link rel=\\"stylesheet\\" href=\\"index.css\\"><title>üì° Tech radars</title><style>body {
          display: flex;
          flex-direction: column;
          align-items: center;
          height: 100vh;
          margin: 0;
      }
      ul {
          list-style-type: none;
          padding: 0px;
      }
      .link {
          color: black;
          text-decoration: none;
      }

      li a {
          min-width: 150px;
          padding: 3px 10px;
          border-radius: 20px;
          display: inline-block;
          font-size: 95%;
      }

      li a:hover {
          background-color: #eaeef8;
      }
      * {
          font-family: \\"Helvetiva, Museo Sans\\", sans-serif;
          color: #333;

          line-height: 1.6333;
      }
      main {
          display: flex;
          justify-content: center;
          height: 100%;
          min-width: 400px;
      }
      .wrapper {
          display: flex;
          flex-wrap: wrap;
          width: 90%;
          flex: 1;
      }
      .tile{
          text-align: center;
          margin: 20px;
      }
      footer {
          text-align: center;
      }</style></head><body><h1>üì° Tech radars</h1><main><div class=\\"wrapper\\"><div class=\\"tile\\"><h2>ios</h2><ul><li><a class=\\"link\\" href=\\"ios/2021-06-11\\">2021-06-11</a></li></ul></div><div class=\\"tile\\"><h2>js</h2><ul><li><a class=\\"link\\" href=\\"js/2021-06-12\\">2021-06-12</a></li></ul></div></div></main><footer><div style=\\"width: 100%; text-align: center; margin: 30px 0 10px 0; font-size: 80%\\"><a href=\\"https://github.com/qiwi/tech-radar\\">Tech-radar generator</a>. Based on <a href=\\"https://github.com/zalando/tech-radar\\">zalando/tech-radar</a>. Boosted with <a href=\\"https://github.com/11ty/eleventy/\\">11ty</a><br><a class=\\"link\\" href=\\"https://github.com/qiwi\\">QIWI ‚ù§Ô∏è Open Source</a></div></footer></body>"
`;

exports[`generate 11ty app generate navigation page from custom templates 1`] = `
"<!DOCTYPE html><head><meta charset=\\"utf-8\\"><link rel=\\"stylesheet\\" href=\\"index.css\\"><title>test</title><style>body {
          display: flex;
          flex-direction: column;
          align-items: center;
          height: 100vh;
          margin: 0;
      }
      ul {
          list-style-type: none;
          padding: 0px;
      }
      .link {
          color: black;
          text-decoration: none;
      }

      li a {
          min-width: 150px;
          padding: 3px 10px;
          border-radius: 20px;
          display: inline-block;
          font-size: 95%;
      }

      li a:hover {
          background-color: #eaeef8;
      }
      * {
          font-family: \\"Helvetiva, Museo Sans\\", sans-serif;
          color: #333;

          line-height: 1.6333;
      }
      main {
          display: flex;
          justify-content: center;
          height: 100%;
          min-width: 400px;
      }
      .wrapper {
          display: flex;
          flex-wrap: wrap;
          width: 90%;
          flex: 1;
      }
      .tile{
          text-align: center;
          margin: 20px;
      }
      footer {
          text-align: center;
      }</style></head><body><h1>test</h1><main><div class=\\"wrapper\\"><div class=\\"tile\\"><h2>.</h2><ul><li><a class=\\"link\\" href=\\"./2021-06-12\\">2021-06-12</a></li><li><a class=\\"link\\" href=\\"./2021-06-11\\">2021-06-11</a></li></ul></div></div></main><footer><div style=\\"width: 100%; text-align: center; margin: 30px 0 10px 0; font-size: 80%\\"><a href=\\"https://github.com/qiwi/tech-radar\\">Tech-radar generator</a>. Based on <a href=\\"https://github.com/zalando/tech-radar\\">zalando/tech-radar</a>. Boosted with <a href=\\"https://github.com/11ty/eleventy/\\">11ty</a><br><a class=\\"link\\" href=\\"https://github.com/qiwi\\">QIWI ‚ù§Ô∏è Open Source</a></div></footer></body>"
`;

exports[`generate 11ty app generate navigation page title 1`] = `
"<!DOCTYPE html><head><meta charset=\\"utf-8\\"><link rel=\\"stylesheet\\" href=\\"index.css\\"><title>title</title><style>body {
          display: flex;
          flex-direction: column;
          align-items: center;
          height: 100vh;
          margin: 0;
      }
      ul {
          list-style-type: none;
          padding: 0px;
      }
      .link {
          color: black;
          text-decoration: none;
      }

      li a {
          min-width: 150px;
          padding: 3px 10px;
          border-radius: 20px;
          display: inline-block;
          font-size: 95%;
      }

      li a:hover {
          background-color: #eaeef8;
      }
      * {
          font-family: \\"Helvetiva, Museo Sans\\", sans-serif;
          color: #333;

          line-height: 1.6333;
      }
      main {
          display: flex;
          justify-content: center;
          height: 100%;
          min-width: 400px;
      }
      .wrapper {
          display: flex;
          flex-wrap: wrap;
          width: 90%;
          flex: 1;
      }
      .tile{
          text-align: center;
          margin: 20px;
      }
      footer {
          text-align: center;
      }</style></head><body><h1>title</h1><main><div class=\\"wrapper\\"><div class=\\"tile\\"><h2>.</h2><ul><li><a class=\\"link\\" href=\\"./2021-06-12\\">2021-06-12</a></li><li><a class=\\"link\\" href=\\"./2021-06-11\\">2021-06-11</a></li></ul></div></div></main><footer><div style=\\"width: 100%; text-align: center; margin: 30px 0 10px 0; font-size: 80%\\"><a href=\\"https://github.com/qiwi/tech-radar\\">Tech-radar generator</a>. Based on <a href=\\"https://github.com/zalando/tech-radar\\">zalando/tech-radar</a>. Boosted with <a href=\\"https://github.com/11ty/eleventy/\\">11ty</a><br><a class=\\"link\\" href=\\"https://github.com/qiwi\\">QIWI ‚ù§Ô∏è Open Source</a></div></footer></body>"
`;
